ta/trusted_key/entry.c

```diff
@@ -0,0 +1,308 @@
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *iv = (uint8_t *)in;
+	uint8_t *tag = (uint8_t *)in + IV_SIZE;
+	uint8_t *enc_key = (uint8_t *)in + IV_SIZE + TAG_SIZE;
+	size_t enc_key_len = 0;
+
+	if (SUB_OVERFLOW(in_sz, (IV_SIZE + TAG_SIZE), &enc_key_len))
+		return TEE_ERROR_SECURITY;
+
+	res = TEE_AEInit(crypto_op, iv, IV_SIZE, TAG_SIZE * 8, 0, 0);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = TEE_AEDecryptFinal(crypto_op, enc_key, enc_key_len, out, out_sz,
+				 tag, TAG_SIZE);
+	if (res != TEE_SUCCESS)
+		res = TEE_ERROR_SECURITY;
+
+	return res;
+}
+
+static TEE_Result huk_crypt(TEE_OperationMode mode, void *in, size_t in_sz,
+			    void *out, size_t *out_sz)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_OperationHandle crypto_op = TEE_HANDLE_NULL;
+	TEE_ObjectHandle hkey = TEE_HANDLE_NULL;
+	uint8_t huk_key[TA_DERIVED_KEY_MAX_SIZE] = { 0 };
+	TEE_Attribute attr = { 0 };
+
+	res = TEE_AllocateOperation(&crypto_op, TEE_ALG_AES_GCM, mode,
+				    sizeof(huk_key) * 8);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = derive_unique_key(huk_key, sizeof(huk_key), NULL, 0);
+	if (res != TEE_SUCCESS) {
+		EMSG("derive_unique_key failed: returned 0x%x\n", res);
+		goto out_op;
+	}
+
+	res = TEE_AllocateTransientObject(TEE_TYPE_AES, sizeof(huk_key) * 8,
+					  &hkey);
+	if (res != TEE_SUCCESS)
+		goto out_op;
+
+	attr.attributeID = TEE_ATTR_SECRET_VALUE;
+	attr.content.ref.buffer = huk_key;
+	attr.content.ref.length = sizeof(huk_key);
+
+	res = TEE_PopulateTransientObject(hkey, &attr, 1);
+	if (res != TEE_SUCCESS)
+		goto out_key;
+
+	res = TEE_SetOperationKey(crypto_op, hkey);
+	if (res != TEE_SUCCESS)
+		goto out_key;
+
+	if (mode == TEE_MODE_ENCRYPT) {
+		res = huk_AE_encrypt(crypto_op, in, in_sz, out, out_sz);
+		if (res != TEE_SUCCESS)
+			EMSG("huk_AE_encrypt failed: returned 0x%x\n", res);
+	} else if (mode == TEE_MODE_DECRYPT) {
+		res = huk_AE_decrypt(crypto_op, in, in_sz, out, out_sz);
+		if (res != TEE_SUCCESS)
+			EMSG("huk_AE_decrypt failed: returned 0x%x\n", res);
+	}
+
+out_key:
+	TEE_FreeTransientObject(hkey);
+out_op:
+	TEE_FreeOperation(crypto_op);
+	memzero_explicit(huk_key, sizeof(huk_key));
+	return res;
+}
+
+static TEE_Result seal_trusted_key(uint32_t types,
+				   TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_SUCCESS;
+	size_t in_sz = 0, out_sz = 0;
+
+	FMSG("Invoked TA_CMD_SEAL");
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				     TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	in_sz = params[0].memref.size;
+	out_sz = params[1].memref.size;
+
+	if (!params[0].memref.buffer || in_sz == 0 || in_sz > MAX_BUF_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+	if (!params[1].memref.buffer || out_sz <= (IV_SIZE + TAG_SIZE) ||
+	    out_sz > MAX_BUF_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if ((in_sz + IV_SIZE + TAG_SIZE) > out_sz)
+		return TEE_ERROR_BAD_PARAMETERS;
```

Comment:
Handling of `out_sz` is a bit unconventional. Why not use the usual scheme with `TEE_ERROR_SHORT_BUFFER` like in `get_devices()`? https://github.com/OP-TEE/optee_os/blob/3.8.0/core/pta/device.c#L34-L67

And same in `unseal_trusted_key()`.

Reply:
Okay, will return `TEE_ERROR_SHORT_BUFFER` here instead.